working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+# Shared Memory Multithreading for Dart
+
+This proposal tries to address two long standing gaps which separate Dart from other more low-level languages:
+
+* Dart developers should be able to utilize multicore capabilities without hitting limitations of the isolate model.
+
+* Interoperability between Dart and native platforms (C/C++, Objective-C/Swift, Java/Kotlin) requires aligning concurrency models between Dart world and native world. The misalignment is not an issue when invoking simple native synchronous APIs from Dart, but it becomes a blocker when:
+
+  * working with APIs pinned to a specific thread (e.g. `@MainActor` APIs in Swift or UI thread APIs on Android) 
+
+  * dealing with APIs which want to call _back_ from native into Dart on an arbitrary thread.
+
+  * using Dart as a container for shared business logic - which can be invoked on an arbitrary thread by surrounding native application. 
+
+  Native code does not _understand_ the concept of isolates. 
+
+The biggest challenge with introducing shared memory multithreading into Dart is avoiding breaking existing code, which is written under the assumption of single-threaded access. We would like to avoid just outright breaking existing code in subtle ways. Consider for a moment a library which is maintains a global cache internally:  
+
+```     dart
+library foo;
+
+class C {
+  final id;
+  C(this.id);
+}
+
+int _nextId;
+final _cache = <int, C>{};
+
+C makeObject() {
+  map[nextId] = C(nextId);
+  nextId++;
+  return map[nextId];
+}
+```
+
+This was a valid way to structure this code in single-threaded Dart, but the same code becomes thread _unsafe_ in the presence of shared memory multithreading.
+
+## Shareable Data
+
+We propose to extend Dart with a shared memory multithreading but provide a clear type-based delineation between two concurrency worlds. **Type of an instance should provide a clear indication whether its fields can be concurrently mutated by another thread.** We are going to call such types _shareable_ and assume that the language provides some way to opt-in into this for classes:
```

Comment:
As a general comment, I think a type based approach is a potentially viable approach, but it may just be impossible to get *right* due on generics.

The idea that some types are inherently "sharable", all subtypes of sharable types are shareable, and all instances of sharable types are shareable, makes reasonable sense for plain classes.
It generalizes to records, which are just multiple plain values.

It's not so simple for functions, and it feels just off to generics.

It's like asynchrony and futures, [one more color](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) that a function can have. 
If you need fully shareable values and functions for something, then any single small non-shareable bit getting into your computation makes the entire thing non-shareable.

And generics are about abstracting over types.
You *can't* reasonably abstract over asynchrony, and you can't abstract over shareability.

The way things are defined here, a class either implements `Shareable` or not, and if so, its state cannot touch non-shareable data, directly or indirectly.
It says that `List<T>` is shareable if `T` is, which I'm not sure is meaningful.
*Maybe* if we say that a class is shareable if its state is, then `List<T>`'s state may contain `T`s, but if everything else is shareable, then the state is shareable if `T` is. But that's an emergent property of the type, not a declared one. It doesn't mesh with doing `implements Shareable`.

What is needed is the ability to have *conditional* shareability. 
Maybe:
```dart
class MyCollection<T> implements Shareable<T> { ... }
```
which abuses notation (`Shareable` isn't generic) to really mean `MyCollection<T>` has `Shareable` as superinterface if and only if `T` is shareable.

That'll be tricky. The "type has superinterface only for some type arguments" was what we dodged when we made extension types not implement `Object` when their representation types does (`extension type E<T>(T _) {}` would have `Object` as superinterface only when `T` did).
It made reasoning quite complicated. (But it wasn't *impossible*.)

Reply:
> It's like asynchrony and futures, [one more color](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) that a function can have.

I had some similar thoughts, but didn't know how to put it into words.

The worry is that introducing another color, or I guess orthogonal color (smell?😆) to to async/non-async, would be a much more invasive change than introducing types.